File: bench/primes.c
// Benchmark 3: Sieve of Eratosthenes (C baseline)
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
int main() {
    int limit = 1000000;
    char *composite = calloc(limit+1, 1);
    for (int p = 2; (long)p*p <= limit; p++)
        if (!composite[p])
            for (int m = p*p; m <= limit; m += p)
                composite[m] = 1;
    int count = 0;
    for (int n = 2; n <= limit; n++)
        if (!composite[n]) count++;
    printf("primes below 1_000_000 = %d\n", count);
    free(composite);
    return 0;
}

File: bench/primes.py
def sieve(limit):
    composite = [False] * (limit + 1)
    p = 2
    while p * p <= limit:
        if not composite[p]:
            for m in range(p*p, limit+1, p):
                composite[m] = True
        p += 1
    return sum(1 for n in range(2, limit+1) if not composite[n])

print(f"primes below 1_000_000 = {sieve(1_000_000)}")
